Fix the problem of index update for the columns that have collection cardinality.

ph1lm opened the pull request:
Fixes #1583

-----

Thank you for contributing to JanusGraph!

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there an issue associated with this PR? Is it referenced in the commit message?
- [x] Does your PR body contain #xyz where xyz is the issue number you are trying to resolve?
- [x] Has your PR been rebased against the latest commit within the target branch (typically `master`)?
- [x] Is your initial contribution a single, squashed commit?

### For code changes:
- [x] Have you written and/or updated unit tests to verify your changes?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](https://www.apache.org/legal/resolved.html#category-a)?
- [ ] If applicable, have you updated the LICENSE.txt file, including the main LICENSE.txt file in the root of this repository?
- [ ] If applicable, have you updated the NOTICE.txt file, including the main NOTICE.txt file found in the root of this repository?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?
- [ ] If this PR is a documentation-only change, have you added a `[skip ci]`
  tag to the first line of your commit message to avoid spending CPU cycles in
  Travis CI when no code, tests, or build configuration are modified?

### Note:
Please ensure that once the PR is submitted, you check Travis CI for build issues and submit an update to your PR as soon as possible.

porunov commented on janusgraph-core/src/main/java/org/janusgraph/diskstorage/indexing/IndexMutation.java:
```diff
@@ -68,22 +78,24 @@ public void resetDelete() {
+    private final Function<IndexEntry, AbstractMap.SimpleEntry<String, Object>> entryConversionFunction =
+            indexEntry -> isCollection(Objects.requireNonNull(indexEntry).field) ?
+                    new AbstractMap.SimpleEntry<>(indexEntry.field, indexEntry.value) :
+                    new AbstractMap.SimpleEntry<>(indexEntry.field, DUMMY);
```
I think we don't need to wrap `indexEntry.field` into `AbstractMap.SimpleEntry`. I.e. we can run like this:
```
private final Function<IndexEntry, Object> entryConversionFunction =
    indexEntry -> isCollection(Objects.requireNonNull(indexEntry).field) ?
        new AbstractMap.SimpleEntry<>(indexEntry.field, indexEntry.value) :
        indexEntry.field;
```

porunov commented on janusgraph-core/src/main/java/org/janusgraph/diskstorage/indexing/IndexMutation.java:
```diff
@@ -68,22 +78,24 @@ public void resetDelete() {
+    }
+
+    private final Function<IndexEntry, AbstractMap.SimpleEntry<String, Object>> entryConversionFunction =
+            indexEntry -> isCollection(Objects.requireNonNull(indexEntry).field) ?
```
Also, could you explain why are you executing `Objects.requireNonNull(indexEntry).field` ?
As for me it is unnecessary because both `Objects.requireNonNull(indexEntry).field` and `indexEntry.field` produces the same `NullPointerException`.

ph1lm commented:
@porunov 
> I think we don't need to wrap indexEntry.field into AbstractMap.SimpleEntry
I just wouldn't feel comfortable if function returns `Object` but you're right it's not needed here.

> Also, could you explain why are you executing Objects.requireNonNull(indexEntry).field
OMG it was auto-generated while converting to lambda-style function...removed.
 
Done with your comments, squashed and re-pushed again.

porunov commented on janusgraph-core/src/main/java/org/janusgraph/diskstorage/indexing/IndexMutation.java:
```diff
@@ -33,19 +34,30 @@
     private boolean isDeleted;
 
-    public IndexMutation(List<IndexEntry> additions, List<IndexEntry> deletions, boolean isNew, boolean isDeleted) {
+    private final Function<IndexEntry, Object> entryConversionFunction =
```
Sorry I forgot also to mention that this function was `static`. I think it should be `private static final Function<IndexEntry, Object>`.

porunov replied:
I see that is uses `isCollection` method, so we can make both `entryConversionFunction` and `isCollection` to be `static`.

porunov replied:
My bad. I see that `isCollection` is uses `storeRetriever` which isn't static.

chupman commented on janusgraph-core/src/main/java/org/janusgraph/diskstorage/indexing/IndexMutation.java:
```diff
@@ -33,19 +34,30 @@
     private boolean isDeleted;
 
-    public IndexMutation(List<IndexEntry> additions, List<IndexEntry> deletions, boolean isNew, boolean isDeleted) {
+    private final Function<IndexEntry, Object> entryConversionFunction =
```
Where is the apply method within this function? Consolidate still expects it to be present. If this doesn't cause a test failure we have a coverage issue we'll need to look into. 

Here's consolidate for reference:
https://github.com/JanusGraph/janusgraph/blob/1e64abc197f35292f0ba55eaa72a492208af3ad3/janusgraph-core/src/main/java/org/janusgraph/diskstorage/Mutation.java#L150-L153

ph1lm replied:
It's a syntax sugar of java 8 which allows to create anonymous single-method-class using lambda expression.

porunov replied:
> Where is the apply method within this function?

It is represented as lambda function. 

Looks like those are flaky tests. I've restarted them

porunov replied:
@chupman Tests passed

chupman replied:
Sorry for the delay, but I had to do some more testing and reading before I was comfortable approving.
